File: Java/Medium/Count Dominant Nodes in a Binary Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
 // TC = O(N)
 // Aux SC = O(logN)
class Solution {
    private int answer;

    public int countDominantNodes(TreeNode root) {
        this.answer = 0;
        dfs(root);
        return this.answer;
    }

    public int dfs(TreeNode node) {
        if (node == null)
            return 1;

        int left = dfs(node.left);
        int right = dfs(node.right);

        if (node.val >= Math.max(left, right))
            this.answer += 1;

        return Collections.max(Arrays.asList(node.val, left, right));
        // return Stream.of(node.val, left, right).mapToInt(Integer::intValue).max().getAsInt();
    }
}

File: Python/Medium/Count Dominant Nodes in a Binary Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def countDominantNodes(self, root: TreeNode | None) -> int:
        ans = 0

        def dfs(n):
            nonlocal ans

            if n is None:
                return 1

            l = dfs(n.left)
            r = dfs(n.right)

            if n.val >= max(l, r):
                ans += 1

            return max(l, r, n.val)

        dfs(root)
        return ans
